Automattic/jetpack
# CLI: Add `jetpack clean`

The `jetpack clean` command allows you to clean your working directory a myriad of ways.

#### Changes proposed in this Pull Request:

* Allows you to run `jetpack clean` which will let you remove:
* - untracked files 
* - node_modules
* - vendor 
* - composer.lock
* - docker files
* - other untracked files

## Running the command 
You can run it by using `jetpack clean`, which will then prompt you for a scope (specific project, project type, or the entire root), and the files that you want to clean. 

## Passthroughs

It also accepts passthroughs:

**Project** 
The project you want to clean:
- all (everything in the monorepo)
- project types - ie, plugins, packages, gh-actions, etc.
- specific projects - ie, plugins/jetpack, packages/abtest, etc

ex. `jetpack clean plugins/jetpack` will then let you choose what files in the `jetpack` plugin that you want to delete. `jetpack clean plugins` will let you choose which files to clean for all of the plugin projects, etc.

**Include**

*alias: 'i'*
What files you want to include:
- untracked
- ignored (any Git-ignored files)
- node_modules
- vendor
- composer.lock
- docker

ex. `jetpack clean plugins node_modules` will nuke all the node_modules folders in every plugin in your `plugins` folder.

You can pass multiple included files using the flag, ie. `jetpack clean plugins -i node_modules vendor`

## Options

`Jetpack clean` can take two options:

**dist**

Running `jetpack clean --dist` will clean all node_modules and vendor folders that might have been generated after running `jetpack build`

**all**

*alias: 'a'*
Running `jetpack clean --all` or `jetpack clean -a` will clean everything in your entire checkout, (except for checked-in composer.lock and docker files). It will clean:

- any untracked files/folders
- all gitignored files
- node_modules
- vendor

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






`jetpack clean` can do quite a lot, so test it any which way you can. Some suggestions for common use cases: 

Check out this branch, then use `jetpack generate` and create a few dummy projects, or add test files to random directories.

- run `jetpack clean`. Follow the prompts to delete the dummy projects from a specific type, or the entire monorepo.
- Try passthroughs. `jetpack clean all` will let you select what file you want to delete from the entire checkout root.
- try nuking_modules a few different ways, ie `jetpack clean` -> then choosing node_modules, trying `jetpack clean all node_modules`, etc. You'll have to run `pnpm install` if you delete the one from root.
- run `jetpack build` and `jetpack install -all` then run `jetpack clean --dist` to clean everything up again. 
- You can run `find -name [your file/folder] -prune` from the repo root to check if the files/folders you wanted were cleaned up properly after running the command.

**Feedback Welcome**

- Is `dist` cleaning everything it should for built files? Is anything missing?
- Is there something being nuked running any of these commands that shouldn't be or that we should make sure gets excluded? 
- Deleting is destructive, so I defaulted to running dry-runs first and requiring confirmation (sometimes twice in the case of `-all`). I'm open to adjusting this if we think it might be too annoying. 
- Naming - very open to feedback on option names/aliases or prompt titles to make things clearer.
